### DISCUSSION TITLE
Refresh / reload read $() and `` expansion output for file picker options reload(live change pick directory) inside fzf reload()

### DISCUSSION THREAD

**Discussion by MikeLemo1** [junegunn/fzf#3673] [Q&A]
I'm trying to do something that might no be possible / easly achieveable and that is turn fzf into a more flexable and dynamic file navigator and manager that will allow for like search toggling between files and directories with a shortcut and between root, home and relative with another (or 3 one per path search if comparmise will be needed)

```bash
  local FILE_PICKER_CMD="fd --hidden"
  set -x
  local SearchPath=~
  local dir
  
dir=$(fd --type file --hidden .  | fzf-tmux --ansi \
    --header "CTRL-T: Switch between Files/Directories CTRL-Y Yank marked" \
    --bind "ctrl-r:execute(echo \"~/.config\" > /tmp/fpicker-dir)" \
    --bind "ctrl-e:execute(echo . > /tmp/fpicker-dir)" \
    --bind "ctrl-t:transform:[[ ! {fzf:prompt} =~ Files ]] &&
      echo \"change-prompt(Files `cat /tmp/fpicker-dir`)+reload(${FILE_PICKER_CMD} --type file . `cat /tmp/fpicker-dir`)\" ||
      echo \"change-prompt(Dirs  `cat /tmp/fpicker-dir`)+reload(${FILE_PICKER_CMD} --type directory . `cat /tmp/fpicker-dir`)\"" \
    --preview "[[ $FZF_PROMPT =~ Files ]] && bat --color=always {} || tree -C {}" \
    --prompt "Files ~" \
    +m)
```
 
and does manage to reload update the picker search dir after quitting fzf and reopening fzf then press ++ctrl+t++ (but not live while in fzf changing fzf search dir using the ++ctrl+e++ and ++ctrl+r++ custome shortcuts)

Am I doing something wrong?

**Comment by LangLangBart** [comment]
> …but not live while in fzf changing fzf search dir using the ++ctrl+e++ and ++ctrl+r++ custome shortcuts

You could try to escape your `cat` command in your `transform` action. I did a very similar mistake, see #3575.

```bash
\$(cat /tmp/fpicker-dir)
# or
\$(</tmp/fpicker-dir)
```

---


Alternatively, you could leverage the `--listen`[^1] flag and use exported functions to send updates
back to `fzf`.


```bash
#!/usr/bin/env bash
set -o errexit -o nounset -o pipefail
# https://www.gnu.org/software/bash/manual/bash.html#The-Set-Builtin

# 'fzf' executes commands with '$SHELL -c'. If your 'SHELL' is 'zsh', this script wouldn't work
# without the line below.
SHELL="$(which bash)"
# Variables are not visible to child processes unless they are exported
export SAVE_DIR="/tmp/fpicker-dir"
export FILE_PICKER_CMD="fd --hidden"
echo "." >"$SAVE_DIR"
# =================== helper functions ====================

# Send a POST request to fzf
send_request_to_fzf() {
	command curl --silent --request POST --data "$1" "localhost:$FZF_PORT"
}

switch_between_dot_and_config() {
	if [[ $FZF_PROMPT =~ config ]]; then
		echo "." >$SAVE_DIR
	else
		echo "$HOME/.config" >$SAVE_DIR
	fi
	if [[ $FZF_PROMPT =~ Files ]]; then
		send_request_to_fzf "change-prompt(Files $(<$SAVE_DIR))"
		$FILE_PICKER_CMD --type file . "$(<$SAVE_DIR)"
	else
		send_request_to_fzf "change-prompt(Dirs  $(<$SAVE_DIR))"
		${FILE_PICKER_CMD} --type directory . "$(<$SAVE_DIR)"
	fi
}

switch_between_files_and_directories() {
	if [[ ! $FZF_PROMPT =~ Files ]]; then
		send_request_to_fzf "change-prompt(Files $(<$SAVE_DIR))"
		$FILE_PICKER_CMD --type file . "$(<$SAVE_DIR)"
	else
		send_request_to_fzf "change-prompt(Dirs  $(<$SAVE_DIR))"
		${FILE_PICKER_CMD} --type directory . "$(<$SAVE_DIR)"
	fi
}

# Functions are not visible to child processes unless they are exported
export -f send_request_to_fzf switch_between_dot_and_config switch_between_files_and_directories

# ====================== check requirements =======================

# https://github.com/junegunn/fzf/commit/5e6788c
min_fzf_version="0.47.0"

user_version=$(fzf --version 2>&1 | grep -Eo '[0-9]+(\.[0-9]+)*' | sed q)

IFS='.' read -ra ver_parts <<<"$user_version"
IFS='.' read -ra threshold_parts <<<"$min_fzf_version"

for i in "${!threshold_parts[@]}"; do
	if ((i >= ${#ver_parts[@]})) || ((ver_parts[i] < threshold_parts[i])); then
		echo "Your 'fzf' version is: '$user_version'"
		echo "The minimum required version is: '$min_fzf_version'."
		exit 1
	elif ((ver_parts[i] > threshold_parts[i])); then
		break
	fi
done

# ===================== let's begin ========================

fzf-tmux \
	--ansi \
	--bind "start:reload:$FILE_PICKER_CMD --type file . $(<$SAVE_DIR)" \
	--bind "ctrl-r:reload:switch_between_dot_and_config" \
	--bind "ctrl-t:reload:switch_between_files_and_directories" \
	--disabled \
	--header "CTRL-R: Switch between '.' and '~/config' | CTRL-T: Switch between Files/Directories" \
	--listen \
	--no-multi \
	--preview $'[[ $FZF_PROMPT =~ Files ]] && bat --color=always {} || tree -C {}' \
	--preview-window 'hidden' \
	--prompt "Files $(<$SAVE_DIR)"
```

EDIT: Updated requirement check

---

> fzf into a more flexible and dynamic file navigator and manager

> [!NOTE]
>  Relevant comment by the maintainer:
>
> _fzf was designed to be an interactive grep rather than a framework for building TUI applications …_
> _We already have nice terminal file managers such as ranger[^2], nnn[^3], and xplr[^4] (which even states that it was inspired by fzf)._
>
> Source:  [PR comment #2440](https://github.com/junegunn/fzf/pull/2440#issuecomment-819136594)


![ranger](https://github.com/junegunn/fzf/assets/92653266/0986899f-ac21-4144-b845-d15884bc62b7)

[^1]: [fzf/CHANGELOG.md 0.36.0 · junegunn/fzf](https://github.com/junegunn/fzf/blob/master/CHANGELOG.md#0360)
[^2]: [Ranger — console file manager with VI key bindings](https://ranger.github.io/)
[^3]: [GitHub - jarun/nnn: n³ The unorthodox terminal file manager](https://github.com/jarun/nnn)
[^4]: [GitHub - sayanarijit/xplr: A hackable, minimal, fast TUI file explorer](https://github.com/sayanarijit/xplr)

**Reply by MikeLemo1** [reply]
Hi I'm prettry much achieving it with the query variable field as it can be accessed and is dynamically from `reload()`
the thing is that I'm dynamically trying to read `cat /tmp/fpicker-dir` in reload which is in a conditional `tranform()` inside double quotes "" inside escaped echoed double quotes \"\" tryed you suggestion like so as well as with escpaed parentasis like so:
```bash
    --bind "ctrl-f:transform:[[ ! {fzf:prompt} =~ Files ]] && 
      echo \"change-prompt(Files >)+reload(${FILE_PICKER_CMD} --type file . \$\(cat /tmp/fpicker-dir\))+transform-header(cat /tmp/fpicker-dir)+transform-query(cat /tmp/fzf-q)\" ||
      echo \"change-prompt(Dirs  >)+reload(${FILE_PICKER_CMD} --type directory . \$\(cat /tmp/fpicker-dir\))+transform-header(cat /tmp/fpicker-dir)+transform-query(cat /tmp/fzf-q)\"" \
```
Which doesn't work 
and here's my current working but somewhat bloated solution:

```bash
    --bind "ctrl-t:transform(echo {fzf:query} > /tmp/fzf-q)+transform-query(cat /tmp/fpicker-dir)+transform:[[ ! {fzf:prompt} =~ Files ]] && 
      echo \"change-prompt(Files >)+reload(${FILE_PICKER_CMD} --type file . {q})+transform-header(cat /tmp/fpicker-dir)+transform-query(cat /tmp/fzf-q)\" ||
      echo \"change-prompt(Dirs  >)+reload(${FILE_PICKER_CMD} --type directory . {q})+transform-header(cat /tmp/fpicker-dir)+transform-query(cat /tmp/fzf-q)\"" \
```

and as for your script I made a file called "fzf_file_picker" and gave it `chmod +x` then put it inside '~/.fzf/bin/' from terminal running `fzf_file_picker` this does seem to run it but hitting ++ctrl+t++ OR ++ctrl+r++ pops an fzf error under query saying 
> [Command failed: switch_between_files_and_directories]

maybe I did something wrong unless you didn't test it?
Also had to update fzf to current 48 from 46

As for the file manager even the vifm I use has an option to use fzf inside it similar to vim so I'm developinng it to find files and change search contexts quicker in bash and in vifm quite like vifm it could just use another column (to be 3 instead of 2 vertical) like in ranger where one in narrower to display previous dire tree and other 2 to be either 2 different directories OR 1 lister and 1 previewer
for some reason I had tentency to prefer vifm over ranger because it's not written in python and from overviewer there and there I understood that ranger is a bit less portable and generally buggy especially with the image previewer in comparison to vifm

Is there a good reason to like ranger / nn better than vifm for nvim vim users?

**Reply by MikeLemo1** [reply]
I also want to add rigpreg toggleable option to the ++ctrl+t++ Dir / Files  query search but not sure if:
`[[ condition ]] && Files option || Dirs options || RipGrep options` would work like intended pressing ++ctrl+t++ 3 times to toggle between all ?

**Reply by LangLangBart** [reply]
> > [Command failed: switch_between_files_and_directories]
> 
> maybe I did something wrong unless you didn't test it?


Thank you for testing it. The script functioned correctly on `macOS`. I suspect the
`FILE_PICKER_CMD` may be misinterpreted. Although using `eval` could potentially solve the issue, it
would be best to first debug it to identify the cause of the failure on your operating system.

```diff
--- a
+++ b
@@ -32,13 +32,16 @@ switch_between_dot_and_config() {
 }
 
 switch_between_files_and_directories() {
-	if [[ ! $FZF_PROMPT =~ Files ]]; then
-		send_request_to_fzf "change-prompt(Files $(<$SAVE_DIR))"
-		$FILE_PICKER_CMD --type file . "$(<$SAVE_DIR)"
-	else
-		send_request_to_fzf "change-prompt(Dirs  $(<$SAVE_DIR))"
-		${FILE_PICKER_CMD} --type directory . "$(<$SAVE_DIR)"
-	fi
+	set -x
+	{
+		if [[ ! $FZF_PROMPT =~ Files ]]; then
+			send_request_to_fzf "change-prompt(Files $(<$SAVE_DIR))"
+			$FILE_PICKER_CMD --type file . "$(<$SAVE_DIR)"
+		else
+			send_request_to_fzf "change-prompt(Dirs  $(<$SAVE_DIR))"
+			${FILE_PICKER_CMD} --type directory . "$(<$SAVE_DIR)"
+		fi
+	} 2>./debug.txt
 }
```

Exemplary Output of 'debug.txt'
```txt
+ [[ ! Dirs  . =~ Files ]]
+ send_request_to_fzf 'change-prompt(Files .)'
+ command curl --silent --request POST --data 'change-prompt(Files .)' localhost:52048
+ fd --hidden --type file . .
```

---

> Is there a good reason to like ranger / nn better than vifm for nvim vim users?


Just a friendly note, since I don't know how aware you already were of other file manager tools.
Whatever works for you is best.

---

> I also want to add rigpreg toggleable option to the ++ctrl+t++ Dir / Files query search but not sure if:
[[ condition ]] && Files option || Dirs options || RipGrep options would work like intended pressing ++ctrl+t++ 3 times to toggle between all ?

If the condition is based on `{fzf:query}`, it might lead to some false positives in an
`if…elif…else` statement. It would be better to use an external file or the `FZF_PROMPT` variable
for the conditional check.

More _ sophisticated_ transformers are possible, as seen in the changelog[^1]. My preference is to use the
`--listen` flag and exported functions for more complex `fzf` scripts. This approach allows more
code to fall under syntax highlighting, auto formatting, and even shellcheck[^2] in my editor.

[^1]: [0.46.0 fzf/CHANGELOG.md](https://github.com/junegunn/fzf/blob/master/CHANGELOG.md#0460)
[^2]: [ShellCheck – shell script analysis tool](https://www.shellcheck.net/)

**Reply by MikeLemo1** [reply]
Well apperantry fzf doesn't even call those functions for some reason and from my understanding `--listen` is quite new and I don't understand the useage of this function as there aren't many good examples that I could find.
and just like the function call's I do in reload with ```` only call `$FZF_PROMPT` once at the beginning and not dynamically like the `{q}` variable which does load "dynamiclly" I think you used this solutions in one of your threads in the past not sure if / how you upgraded it since.
